include/simics/dmllib.h

```diff
@@ -2560,12 +2593,144 @@ UNUSED static void _free_table(ht_int_table_t *table) {
     char msg[512];
-    snprintf(msg, 512,
-             "Recursive call to memoized method %s. "
-             "This is considered undefined behavior.", name);
+    vsnprintf(msg, 512, format, va);
+    va_end(va);
     VT_critical_error(msg, msg);
 }
 
+UNUSED static void _memoized_recursion(const char *name) {
+    _signal_critical_error("Recursive call to memoized method %s. "
+                           "This is considered undefined behavior.", name);
+}
+
+typedef struct {
+    bool valid;
+    bool is_bank;
+    bool is_array;
+    // uint32 array_size;
+    ptrdiff_t port_obj_offset;
+    const char *portname;
+} _attr_parent_obj_proxy_info_t;
+
+typedef struct {
+    const char *name;
+    const char *type;
+    const char *doc;
+    conf_class_t *parent_obj_class;
+    _attr_parent_obj_proxy_info_t proxy_info;
+    attr_attr_t flags;
+    uint32 start_dim;
+    bool readable;
+    bool writable;
+    bool generated;
+    bool allow_cutoff;
+} _dml_attr_conf_info_t;
+
+typedef struct {
+    const _id_info_t *id_info;
+    void *vtable;
+    uint32 num;
+    bool allow_cutoff;
+} _dml_attr_getset_info_t;
+
+UNUSED static void _DML_register_attributes(
+    conf_class_t *dev_class, const _id_info_t *id_info_array,
+    const _vtable_list_t *attribute_vtables, _each_in_t sequence,
+    _dml_attr_conf_info_t (*get_attribute_info)(_traitref_t),
+    attr_value_t (*get_attr)(void *, conf_object_t *, attr_value_t *),
+    set_error_t (*set_attr)(void *, conf_object_t *, attr_value_t *,
+                            attr_value_t *),
+    attr_value_t (*get_portobj_attr)(void *, conf_object_t *, attr_value_t *),
+    set_error_t (*set_portobj_attr)(void *, conf_object_t *, attr_value_t *,
+                            attr_value_t *)) {
+    for (uint32 i = 0; i < sequence.num; ++i) {
+        _vtable_list_t list = attribute_vtables[i];
+        uint64 num = list.num / sequence.array_size;
+        uint64 start = num * sequence.array_idx;
+        _traitref_t traitref = { list.vtable,
+                                 { .id = list.id, .encoded_index = start }};
+        _dml_attr_conf_info_t attr_info = get_attribute_info(traitref);
+        if (!attr_info.generated) {
+            continue;
+        }
+
+        _dml_attr_getset_info_t attr_getset_info = {
+            &id_info_array[list.id],
+            traitref.trait,
+            num,
+            attr_info.allow_cutoff
+        };
+        _dml_attr_getset_info_t *attr_get_info = NULL;
+        if (attr_info.readable) {
+            attr_get_info = MM_MALLOC(1, _dml_attr_getset_info_t);
+            *attr_get_info = attr_getset_info;
+        }
+        _dml_attr_getset_info_t *attr_set_info = NULL;
+        if (attr_info.writable) {
+            attr_set_info = MM_MALLOC(1, _dml_attr_getset_info_t);
+            *attr_set_info = attr_getset_info;
+        }
+
+        char type[1024];
+        char tmp_type[1024];
+        int written = snprintf(type, 1024, "%s", attr_info.type);
```

Comment:
This should just use `strbuf`. I was less acquainted with it at the time I wrote this code.

Reply:
Eh, actually, `strbuf` isn't as clean as one would like since the loop doesn't simply append upon the string. Still, getting rid of the 1024 char cap would be nice. The code would look something like this:
```
        strbuf_t type = sb_newf("%s", attr_info.type)
        _id_info_t id_info = id_info_array[list.id];
        for (int64 i = (int64)id_info.dimensions - 1; i >= attr_info.start_dim;
             --i) {
            char *tmp_type = sb_detach(&type);
            sb_addfmt(&type, attr_info.allow_cutoff ? "[%s{0:%d}]" : "[%s{%d}]",
                    tmp_type, id_info.dimsizes[i]);
            MM_FREE(tmp_type);
        }
```

Reply:
`strbuf` now used.